### ISSUE TITLE
`rtol` argument has no effect in `SparsePauliOp.simplify(... rtol=...)`

### ISSUE THREAD

**Issue by aeddins-ibm** [Qiskit/qiskit#13383]
### Environment

- **Qiskit version**: 5d534f2c
- **Python version**: 3.11
- **Operating system**: macos

### What is happening?

`SparsePauliOp.simplify()` has an argument `rtol` that it passes to calls of the form `np.isclose(a, b, rtol=rtol)`. The docs for `isclose` explain that `rtol` is multiplied by `b` to compute the total tolerance. However, in `simplify()`, `b` is always zero, so `rtol` has no effect on the result of `isclose`, and thus no effect on the result of `simplify`:
https://github.com/Qiskit/qiskit/blob/91e83eff3efb25a4f51141a5bdcbfb4794e09d07/qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py#L475-L500

### How can we reproduce the issue?

Compare the code above to this note from the `isclose` docs:

> Be sure to select atol for the use case at hand, especially for defining the threshold below which a non-zero value in a will be considered “close” to a very small or zero value in b.

### What should happen?

Probably just remove the `rtol` argument? 

Or else define and implement some new expected behavior for it?

**Comment by jakelishman** [comment]
Thanks, yeah - I was confused when I saw it too. Our new `SparseObservable` just has a single `tol` argument, and probably we ought to transition `SparsePauliOp` to the same.

**Comment by ains-validator** [comment]
@validate please write a validation plan for this issue.

**Comment by ains-validator** [comment]
**Validator**

On it — working on branch `pocket/13383`.

```
oc logs -f job/pocket-qiskit-qiskit-13383-1787220254 -n hub
```

**Comment by ains-validator** [comment]
**Validator** — blocked, no plan written.

I did not write a validation plan for this issue. The tool that defines what a validation plan is, `validation_plan_schema` on the **hub-db** MCP server, is not exposed by the server, and both my instructions and the installed `validation-plan` skill say to stop rather than reconstruct the sections from memory.

### What I tried

The hub itself is up and answering — this is not a VPN or connectivity failure:

- `repos` returned all seven scraped repositories, including `Qiskit/qiskit` (`repo_id: 3`).
- `skills` returned the six per-repo skills, including `qiskit-change-validation`.

The missing piece is one specific tool. I checked past my own tool list, straight against the server at `https://hub.vpc-int.res.ibm.com/mcp`:

| Call | Result |
| --- | --- |
| `validation_plan_schema` | `No such tool available` |
| `initialize` | `200`, `serverInfo: {"name": "ains-validation-hub", "version": "3.4.7"}` |
| `tools/list` | exactly six tools: `schema`, `query`, `repos`, `search`, `validations`, `skills` |
| `prompts/list` | `{"prompts": []}` |
| `resources/list` | `{"resources": []}` |
| `resources/templates/list` | `{"resourceTemplates": []}` |

So the shape is not merely missing from my tool namespace — it is not published on the hub as a tool, a prompt, or a resource. The per-repo skills that *are* there describe how a repo's maintainers validate a change (which harness, what evidence); none of them is the definition of a plan's sections, and substituting one for that definition would be the same error as writing the sections from memory.

### Why I did not just write it anyway

The section list is the part I am least able to check. An ungrounded plan and a grounded one are indistinguishable on the page: a plan with five of the required sections reads exactly like a plan with all of them, and nothing in the document would record which definition it had followed. The hub was made the single owner of that shape precisely so a plan that followed the tool was current when it was written. With the tool absent there is nothing to follow, so the honest output is this note rather than a document that looks complete.

### What would unblock this

Either of these, and I can produce the plan in one pass:

1. `validation_plan_schema` restored on the hub-db MCP server (it is absent from `tools/list` at v3.4.7 — this looks like a server-side regression or a version skew between the image's expectations and the deployed hub, not a client misconfiguration), or
2. the section list, in order, with what belongs under each and the cross-section rules, from whatever is now canonical for it.

The grounding work for this particular issue is not the blocker — `Qiskit/qiskit` is `repo_id: 3` in the hub, `schema`/`query`/`search`/`validations` all answer, and `qiskit-change-validation` is installable. Only the shape is missing.

### Scope note, for whoever picks this up

Nothing here touches the substance of the report. The `rtol`/`atol` behaviour in `SparsePauliOp.simplify` is untouched, no code was changed, and no branch was pushed.

**Comment by ains-validator** [comment]
@validate please build a validation plan for this issue.

**Comment by ains-validator** [comment]
**Validator** — validation plan.

Shape from `validation_plan_schema` (hub-db, v1, `validation-plan`); the six sections below are its sections in its order. Grounded against `Qiskit/qiskit` = `repo_id 3` in the hub, and against the working tree at `3fa5a67`. Repo skill `qiskit-change-validation` installed and followed for harness selection. No code changed, no PR opened.

---

## 1. What is being validated

A change to `SparsePauliOp.simplify` in `qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py`. The method accepts `rtol` and threads it into four `numpy.isclose(x, 0, atol=atol, rtol=rtol)` calls; because the reference argument is `0`, the `rtol * |b|` term of the `isclose` predicate is always `0`, so only `atol` decides which coefficients are dropped. `rtol` is accepted, documented as *"relative tolerance for checking if coefficients are zero"*, and ignored. Still true at `3fa5a67`:

```
$ grep -n "rtol" qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py
464:    def simplify(self, atol: float | None = None, rtol: float | None = None) -> SparsePauliOp:
...
508:                np.isclose([to_complex(x) for x in self.coeffs], 0, atol=atol, rtol=rtol)
511:            non_zero = np.logical_not(np.isclose(self.coeffs, 0, atol=atol, rtol=rtol))
516:                [np.isclose(to_complex(coeff), 0, atol=atol, rtol=rtol) for coeff in coeffs]
519:            is_zero = np.isclose(coeffs, 0, atol=atol, rtol=rtol)
```

**The scope is already settled, and the plan validates the settled scope.** @jakelishman ruled it out in #15963 on 2026-04-07: *"We can't actually merge this PR that changes the meaning of the `rtol` argument to anything, because that would be a logically breaking change in the API of the function... the solution to #13383 that's available to us within stability guarantees is just to document/comment that the option is meaningless."* So the change under validation is **documentation and comments only, with runtime behaviour bit-identical**. A plan that measured whether a new `rtol` semantics were correct would be validating a change the repo has already declined.

Two adjacent uses in the same file must be left alone, and the plan measures that they are. `is_unitary` compares against `1.0` at line 459, where `rtol` is genuinely meaningful; `from_operator` computes `tol = max((atol, rtol))` at line 787, a third and different semantics. Only `simplify` is broken.

## 2. Population

Three populations, each the result of a query actually run. Row counts are as returned; caps are stated and none was hit.

**P1 — the tolerance-semantics history of this exact file.** The direct precedent: every PR touching `sparse_pauli_op.py` that argues about `rtol`/`atol`/tolerance. **11 rows** (cap `LIMIT 40`, not reached).

```sql
SELECT i.number, i.title, i.state, i.merged, i.additions+i.deletions AS diff_lines,
       length(i.data->>'body') AS body_chars, i.ai_authored
FROM issues i
WHERE i.repo_id=3 AND i.is_pr
  AND EXISTS (SELECT 1 FROM jsonb_array_elements(i.data->'files'->'nodes') f
              WHERE f->>'path' = 'qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py')
  AND (i.title ~* 'rtol|atol|toleran' OR i.data->>'body' ~* 'rtol|atol|toleran')
ORDER BY i.number DESC LIMIT 40;
```

| PR | State | Diff lines | Body chars | AI-authored |
|---|---|---|---|---|
| #16127 Document that `rtol` has no effect in `SparsePauliOp.simplify` | CLOSED | 9 | 3108 | false |
| #15963 Fix `SparsePauliOp.simplify` ignoring `rtol` for numeric coefficients | CLOSED | 74 | 1503 | **true** |
| #15029 Fix `is_unitary()` to respect input tolerance (backport) | MERGED | 48 | 1110 | false |
| #14764 Fix simplify to sum duplicates before applying tolerance (backport) | MERGED | 53 | 140 | false |
| #14676 Fix `is_unitary()` to respect input tolerance values | MERGED | 48 | 1037 | false |
| #14564 Fix: Respect tolerance in `is_unitary()` | CLOSED | 82 | 81 | false |
| #14425 Fix simplify to sum duplicates before applying tolerance | MERGED | 53 | 556 | false |
| #8281 add `atol` parameter to `SparsePauliOp.equiv` | MERGED | 20 | 603 | false |
| #7598 Expose cutoff tolerances in `Z2Symmetries` | MERGED | 142 | 1433 | false |
| #5617 quantum_info.operator mixin interface refactor | MERGED | 2688 | 2383 | false |
| #5407 Convert tolerances metaclass to a mixin class | MERGED | 231 | 752 | false |

Two of these eleven targeted this issue directly and neither merged. #15963 tried to give `rtol` meaning and was closed on the stability-guarantee argument quoted above. #16127 did exactly the documentation-only change this plan describes and was closed for **process, not content** — *"This is not a statement that this PR is necessarily unmergeable or incorrect, but on balance of probability, it is not worth the maintainer time to make an accurate determination."* That is why §3 measures diff economy and body length alongside correctness: on this issue, those are what the record shows decides the outcome.

**P2 — outcome baseline for a documentation-only change in this subsystem.** **119 rows** (107 merged, 12 not).

```sql
SELECT i.merged, count(*) FROM issues i
JOIN pr_types t ON t.repo_id=3 AND t.number=i.number
WHERE i.repo_id=3 AND i.is_pr AND t.primary_type='docs'
  AND EXISTS (SELECT 1 FROM jsonb_array_elements(i.data->'files'->'nodes') f
              WHERE f->>'path' LIKE 'qiskit/quantum_info/%')
GROUP BY 1;
-- false | 12
-- true  | 107
```

**P3 — evidence-carrying baseline for this file.** **100 rows**: merged PRs touching `sparse_pauli_op.py`.

```sql
SELECT count(*) FROM issues i
WHERE i.repo_id=3 AND i.is_pr AND i.merged
  AND EXISTS (SELECT 1 FROM jsonb_array_elements(i.data->'files'->'nodes') f
              WHERE f->>'path' = 'qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py');
-- 100
```

## 3. Measures

Six measures. M2, M3 and M5 read straight out of the hub; M1, M4 and M6 come from the repo and its harnesses, each with the command that produces the number.

| # | What is counted | Where the number comes from | Better |
|---|---|---|---|
| M1 | Coefficient sets whose `simplify` output differs from `3fa5a67` across a swept `rtol`, `atol` held fixed | `stestr run -n test.python.quantum_info.operators.symplectic.test_sparse_pauli_op`, plus a scratch sweep over float, complex and `ParameterExpression` coeffs | **lower** — 0 is the only acceptable value |
| M2 | Lines changed (`additions + deletions`) | `issues.additions`, `issues.deletions` | **lower** |
| M3 | PR body length in characters | `length(issues.data->>'body')` | **lower** |
| M4 | Exit codes of the two gates the skill maps to a docstring change: `tox -e docs` (`sphinx-build -W`, warnings are errors) and `tox -e lint` | local run; CI jobs `Documentation` and `Lint`, both in the `needs:` of `Finalize` in `.github/workflows/branch-protection.yml` | 0 for both |
| M5 | Release-note files added under `releasenotes/notes/` | `data->'files'->'nodes'` path match | **0** — see baseline; the skill says drop the note for docs-only changes |
| M6 | `np.isclose(..., 0, ...)` sites inside `simplify` left unmodified, and `rtol` uses elsewhere in the file left unmodified | `grep -n "isclose" ... \| sed -n '500,525p'`; `git diff --stat` | 4 of 4, and 0 lines touched outside `simplify`'s docstring/comments |

M1 is the load-bearing one and is cheap precisely because the intended change is documentation: any nonzero M1 means the "documentation-only" claim is false. The skill's §2 rule — *prove the test is load-bearing, confirm it fails before your fix* — inverts here. There is nothing to fail. What must be shown instead is that a test sweeping `rtol` **passes identically before and after**, which is the evidence that the docstring now tells the truth.

## 4. Baseline

**M1 — behaviour drift: 0 by construction at `3fa5a67`, and untested.** No test in the file passes `rtol` to `simplify`:

```
$ grep -c "def test" test/python/quantum_info/operators/symplectic/test_sparse_pauli_op.py
94
$ grep -n "simplify(.*rtol" test/python/quantum_info/operators/symplectic/test_sparse_pauli_op.py
(no matches)
$ grep -n "is_unitary(.*rtol" test/python/quantum_info/operators/symplectic/test_sparse_pauli_op.py
1453:            op.is_unitary(atol=1e-5, rtol=1e-3),
1459:            op.is_unitary(atol=1e-10, rtol=1e-10),
```

94 test methods, 26 mentions of `simplify`, and **zero** assertions about `rtol` in `simplify`. The two `rtol` assertions that exist are on `is_unitary`, the method where it works — so the suite as it stands would not notice `rtol` in `simplify` acquiring or losing meaning.

**M2 / M3 / M5 — P2, split by outcome.**

```sql
WITH pop AS (
  SELECT i.number, i.merged, i.additions, i.deletions, i.changed_files,
         length(i.data->>'body') AS body_chars,
         EXISTS (SELECT 1 FROM jsonb_array_elements(i.data->'files'->'nodes') f WHERE f->>'path' LIKE 'releasenotes/notes/%') AS has_note,
         EXISTS (SELECT 1 FROM jsonb_array_elements(i.data->'files'->'nodes') f WHERE f->>'path' LIKE 'test/%') AS has_test
  FROM issues i JOIN pr_types t ON t.repo_id=3 AND t.number=i.number
  WHERE i.repo_id=3 AND i.is_pr AND t.primary_type='docs'
    AND EXISTS (SELECT 1 FROM jsonb_array_elements(i.data->'files'->'nodes') f WHERE f->>'path' LIKE 'qiskit/quantum_info/%')
)
SELECT merged, count(*) AS n,
  percentile_disc(0.5) WITHIN GROUP (ORDER BY additions+deletions) AS median_diff_lines,
  max(additions+deletions) AS max_diff_lines,
  percentile_disc(0.5) WITHIN GROUP (ORDER BY changed_files) AS median_files,
  percentile_disc(0.5) WITHIN GROUP (ORDER BY body_chars) AS median_body_chars,
  count(*) FILTER (WHERE has_note) AS with_release_note,
  count(*) FILTER (WHERE has_test) AS with_test
FROM pop GROUP BY merged;
```

| merged | n | median diff lines | max diff lines | median files | median body chars | with release note | with test |
|---|---|---|---|---|---|---|---|
| true | 107 | **20** | 3619 | 3 | **683** | 15 | 14 |
| false | 12 | 45 | 2875 | 4 | 617 | 3 | 3 |

A merged docs change in `quantum_info` is 20 lines and a 683-character body. #16127 was 9 lines — comfortably inside — and a **3108-character** body, 4.5× the merged median. Its diff was never the problem.

**M3 context — the attention signal, since that is the stated reason both attempts died.**

```sql
SELECT ai_authored, count(*) AS prs, count(*) FILTER (WHERE merged) AS merged,
       round(100.0*count(*) FILTER (WHERE merged)/count(*),1) AS merge_pct,
       percentile_disc(0.5) WITHIN GROUP (ORDER BY length(data->>'body')) AS median_body_chars
FROM issues WHERE repo_id=3 AND is_pr AND state<>'OPEN' GROUP BY 1;
-- false | 10696 | 9009 | 84.2 |  855
-- true  |    29 |   13 | 44.8 | 1292
```

Narrowed to documentation-primary PRs touching `qiskit/`:

```sql
WITH docs AS (
  SELECT i.number, i.merged, i.ai_authored, length(i.data->>'body') AS body_chars
  FROM issues i JOIN pr_types t ON t.repo_id=3 AND t.number=i.number
  WHERE i.repo_id=3 AND i.is_pr AND i.state<>'OPEN' AND t.primary_type='docs'
    AND EXISTS (SELECT 1 FROM jsonb_array_elements(i.data->'files'->'nodes') f WHERE f->>'path' LIKE 'qiskit/%')
)
SELECT ai_authored, count(*) AS prs, count(*) FILTER (WHERE merged) AS merged,
       round(100.0*count(*) FILTER (WHERE merged)/count(*),1) AS merge_pct,
       percentile_disc(0.5) WITHIN GROUP (ORDER BY body_chars) AS median_body_chars
FROM docs GROUP BY 1;
-- false | 919 | 843 | 91.7 | 737
-- true  |   4 |   0 |  0.0 | 575
```

**0 of 4** AI-authored documentation PRs against `qiskit/` have merged, against 843 of 919 otherwise. n=4 is too small to be a rate, but as a description of the record it is exact, and it is the record this issue's two closures are part of.

**M4 — the gates exist and run.** From the last 3000 workflow runs on `repo_id 3`:

```sql
SELECT data->>'name' AS workflow, count(*) AS runs
FROM (SELECT data FROM workflow_runs WHERE repo_id=3 ORDER BY id DESC LIMIT 3000) s
GROUP BY 1 ORDER BY runs DESC LIMIT 15;
-- Build wheels 973 | Coverage report 895 | Branch Protection 747 | Update caches 109
-- Documentation 77 | Backport metadata 34 | Slow nightly tests 32 | Nightly Main tests 32 ...
```

Capped at the 3000 most recent runs, stated because the unaggregated query over all ~322k rows times out on the hub (`canceling statement due to statement timeout`). The cap makes these recent-activity counts, not lifetime totals; `Documentation` and `Branch Protection` both being present is all M4 needs from them.

**M5 — release note.** 15 of 107 merged docs-only `quantum_info` PRs shipped one, from the P2 table above; the skill's §3 records reviewers asking for the note to be dropped on docs-only changes (#16389, #15700, #16437). Baseline expectation for this change: none.

**M6 — 4 `isclose` sites, unchanged.**

```
$ sed -n '500,525p' qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py | grep -c "isclose"
4
```

**Evidence density on this file, for calibration.** Of the 100 merged PRs in P3:

```sql
WITH pop AS (
  SELECT i.number, i.data, t.primary_type FROM issues i
  LEFT JOIN pr_types t ON t.repo_id=3 AND t.number=i.number
  WHERE i.repo_id=3 AND i.is_pr AND i.merged
    AND EXISTS (SELECT 1 FROM jsonb_array_elements(i.data->'files'->'nodes') f
                WHERE f->>'path' = 'qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py')
), flags AS (
  SELECT p.number, p.primary_type,
    EXISTS (SELECT 1 FROM jsonb_array_elements(p.data->'files'->'nodes') f WHERE f->>'path' LIKE 'test/%') AS has_test,
    EXISTS (SELECT 1 FROM jsonb_array_elements(p.data->'files'->'nodes') f WHERE f->>'path' LIKE 'test/python/quantum_info/%') AS has_qi_test,
    EXISTS (SELECT 1 FROM jsonb_array_elements(p.data->'files'->'nodes') f WHERE f->>'path' LIKE 'releasenotes/notes/%') AS has_note
  FROM pop p
)
SELECT count(*) AS n, count(*) FILTER (WHERE has_test) AS with_any_test,
  count(*) FILTER (WHERE has_qi_test) AS with_quantum_info_test,
  count(*) FILTER (WHERE has_note) AS with_release_note,
  count(*) FILTER (WHERE primary_type='fix') AS fixes,
  count(*) FILTER (WHERE primary_type='fix' AND has_test) AS fixes_with_test
FROM flags;
-- 100 | 64 | 53 | 55 | 29 | 26
```

26 of the 29 merged fixes to this file shipped a test. A behavioural fix here is expected to carry one; this change is not behavioural, which is exactly the claim M1 has to establish.

## 5. Pass condition

Written before any result is seen. All six must hold.

1. **M1 = 0.** For every coefficient set in the sweep — real, complex, and `object`-dtype `ParameterExpression` — `simplify(atol=a, rtol=r)` returns output identical to `3fa5a67` for all swept `r` at fixed `a`. Identical means equal `paulis` and `coeffs`, compared with `assertEqual` on the operators, not `assert_allclose`. A single differing pair fails the whole plan: it means the change is behavioural and falls under the #15963 ruling.
2. **M2 ≤ 20 lines**, the merged-docs median, across ≤ 2 files. Above 45 — the unmerged median — fails.
3. **M3 ≤ 683 characters** of PR body, the merged-docs median. #16127's 3108 is the anti-target. One sentence of what, one of why, the `Fix #13383` line, the filled-in AI/LLM disclosure section of `.github/PULL_REQUEST_TEMPLATE.md`. Deleting that template blocks review outright (#16604, #16157).
4. **M4: `tox -e docs` and `tox -e lint` both exit 0**, output pasted. `sphinx-build -W` means a malformed docstring is a failure, not a warning, and the rendered docstring must be read after building, not assumed.
5. **M5 = 0 release-note files.**
6. **M6: 4 of 4 `isclose` sites unchanged, and `git diff` touches nothing outside `simplify`'s docstring and comments.** In particular `is_unitary` (line 459, `rtol` against `1.0`) and `from_operator` (line 787, `max(atol, rtol)`) are untouched, and `test_sparse_pauli_op.py:1453,1459` still pass.

A pass means: the docstring now says what the code does, the code does what it did, and the change is small enough and quiet enough to be worth a maintainer's attention. It does not mean the PR merges — that is a human decision this plan cannot make and should not pretend to predict.

## 6. What this will not tell you

- **The evaluation map has nothing for these paths.** `validations` for `sparse_pauli_op.py` and `test_sparse_pauli_op.py` returns both under `unmapped`; the only checks that resolved — `Miri` (`cargo miri test`) and `Coverage` — came from the third, hypothetical release-note path, are marked `"verified": "unverified"`, and are irrelevant to a Python docstring. So M4's harness selection rests on the installed repo skill and `branch-protection.yml`, not on the hub's derived map. The map's silence here is visible rather than assumed, which is the most it can offer.
- **Precedent evidence on this file is thin.** Only 13 of the 100 merged PRs in P3 carry any extracted evaluation at all (44 result rows; 42 of them, across 11 PRs, resolve to a harness, and that harness is `"custom script"` in every case). There is no asv, stestr or cargo evidence recorded for this file to compare against.

  ```sql
  WITH pop AS (SELECT i.number FROM issues i WHERE i.repo_id=3 AND i.is_pr AND i.merged
    AND EXISTS (SELECT 1 FROM jsonb_array_elements(i.data->'files'->'nodes') f
                WHERE f->>'path' = 'qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py'))
  SELECT count(DISTINCT p.number) AS prs_with_any_evaluation, count(*) AS evaluation_rows
  FROM pop p JOIN evaluation_results er ON er.repo_id=3 AND er.number=p.number;
  -- 13 | 44
  ```

- **M1 cannot prove the docstring is *correct*, only that behaviour did not move.** A docstring that is newly truthful about `rtol` and newly wrong about something else passes every measure here. Nothing but reading it catches that.
- **M3 is a proxy for a judgement, and a weak one.** #16127 was closed over maintainer attention, not character count; a 600-character body written without understanding would score identically to one written with it. The `ai_authored` numbers are a classifier's output over n=29 repo-wide and n=4 in the docs slice — too small to be a rate. They describe the record, they do not predict a reviewer.
- **Sampling bias in P2.** `primary_type` is classifier-assigned (`pr_types.classifier`), and P2 is docs-*primary*, so a PR that fixed code and documented it lands elsewhere. The 107/12 split is also survivorship-shaped: PRs still open are excluded by `state<>'OPEN'`.
- **The M4 workflow counts are capped at the 3000 most recent runs** because the full aggregate times out. They establish that `Documentation` and `Branch Protection` run; they are not lifetime frequencies, and nothing in the plan should be read as one.
- **Blind to the follow-up.** @jakelishman's first comment on this issue notes `SparseObservable` takes a single `tol` and *"probably we ought to transition `SparsePauliOp` to the same"* — confirmed in the tree: `crates/quantum_info/src/sparse_observable/mod.rs:3458` is `fn simplify(&self, tol: f64)`, as is `pauli_lindblad_map_class.rs:1803`. That migration is a deprecation cycle, needs a release note and `assertWarns(DeprecationWarning)` tests, and is out of scope under the #15963 ruling. Whether it ever happens is **unknown**; what would answer it is a deprecation-policy decision recorded on this issue or a `Changelog: Deprecation`-labelled PR against this file, neither of which exists in the hub today.
- **Not measured: whether the docs-only outcome is the right one.** The plan takes the maintainer ruling as settled and validates against it. If the team reopens the API question, every measure above needs rewriting, starting with M1 — which would invert from "must be 0" to "must be nonzero, in exactly these cases."
